### PR TITLE
More int types

### LINKED ISSUE #127: Implement Int256
Needs #124 

It holds 4 u64 values.

We add an Int64 trait with a static From method that takes an u64.
We evaluate operators to calls to the helpers trait or perhaps we evaluate them directly.

Example operator implementation (the direct way):
```k
   syntax Expression ::= int256ToInt(Expression)  [strict]
   syntax Expression ::= int256ToInt(Expression, Expression, Expression, Expression)  [seqstrict]
   syntax Expression ::= intToInt256(Expression)  [seqstrict]
   syntax KResult ::= wrappedInt(Int)


   rule ptrValue(_, struct(Int256, _)) #as A + ptrValue(_, struct(Int256, _)) #as B
       => intToInt256(int256ToInt(A.f1, a.f2, a.f3, a.f4) + int256ToInt(B.f1, B.f2, B.f3, B.f4))

   rule int256ToInt(V:PtrValue) => int256ToInt(V.x1, V.x2, V.x3, V.x4)
   rule int256ToInt(ptrValue(_, i64(V1)), ...) => wrappedInt(MintToInt(V1) << (64 *Int 3) |Int ...)

   rule intToInt256(wrappedInt(I:Int)) => newStruct(int256Type, I & ((1 <<Int 64) - 1) << (3 *Int 64), I & ((1 <<Int 64) - 1) << (3 *Int 64), I & ((2 <<Int 64) - 1) << (64), I & ((1 <<Int 64) - 1))
   rule wrappedInt(I) + wrappedInt(J) => wrappedInt(I +Int J)
```

### PR DESCRIPTION
**Pull request by virgil-serbanuta**
Fixes #127

### CODE REVIEW

**Review by ACassimiro — APPROVED** [review]
Just for clarification, this modification now allows us to use unsupported types in our contracts, right? And the types will be handled by K as normal integers, without any special handling such as representing them using structs or tuples.

It seems clear to me, and I have no problems with it for our demos.

File: tests/preprocessing/casts_int.rs [hunk 1]
```diff
@@ -71,7 +89,10 @@ pub const I64_05: i64 = 10_i32;
 pub const I64_06: i64 = 10_u32;
 pub const I64_07: i64 = 10_i64;
 pub const I64_08: i64 = 10_u64;
-pub const I64_09: i64 = 10;
+pub const I64_09: i64 = 10_u128;
+pub const I64_10: i64 = 10_u160;
+pub const I64_11: i64 = 10_u256;
```
Comment (yanliu18) [inline tests/preprocessing/casts_int.rs:94]: Is this a legal syntax in Rust?

**Review by yanliu18 — CHANGES_REQUESTED** [review]
I did some research and confirmed that `u256` is not a valid type in Rust, unless you are using some external libraries.
And the [documentation in MultiverseX](https://docs.multiversx.com/developers/data/simple-values/#arbitrary-width-big-numbers) contract uses `biguint` and `bigint` to represent large numbers exceeding `u64`.
The rust builtin implementation of `biguint` can be referenced [here](https://docs.rs/num-bigint/latest/num_bigint/struct.BigUint.html).

Although, we might want to take shortcuts for Devcon demo, we should still maintain rust contract legal. What do you think?

File: tests/preprocessing/casts_int.rs (hunk 1 above)
Comment (virgil-serbanuta) [reply tests/preprocessing/casts_int.rs:94]: No. This is a shortcut that I have taken for the demo. It saves at least a day of development, if not more. I should have added some comments about this in the K code, I will do that now.

**Comment by virgil-serbanuta** [comment]
> Just for clarification, this modification now allows us to use unsupported types in our contracts, right? And the types will be handled by K as normal integers, without any special handling such as representing them using structs or tuples.
> 
> It seems clear to me, and I have no problems with it for our demos.

Yes. This is not a good solution in general, but it saves time for the demo.

**Comment by virgil-serbanuta** [comment]
> I did some research and confirmed that `u256` is not a valid type in Rust, unless you are using some external libraries. And the [documentation in MultiverseX](https://docs.multiversx.com/developers/data/simple-values/#arbitrary-width-big-numbers) contract uses `biguint` and `bigint` to represent large numbers exceeding `u64`. The rust builtin implementation of `biguint` can be referenced [here](https://docs.rs/num-bigint/latest/num_bigint/struct.BigUint.html).

I know. There are two issues here:
1. ULM works with `u256` while `BigUint` is either `u800` or has unlimited size, depending on the context. I don't think we can just replace one with the other.
2. Implementing BigUint takes time (or have a similar implementation for u256). It's true that we would mostly be copy-pasting things from the original MX semantics, but even that takes time.

> 
> Although, we might want to take shortcuts for Devcon demo, we should still maintain rust contract legal. What do you think?

I mostly agree. My original plan was to implement a U256 type as a struct in Rust (for which we would have to fake implementations for operators somehow, and we would need to make encoding and decoding work for structs; we would also have a choice between making `impl` work for structs or using static functions for most things we need - all of these are easy-ish, but they take time, and we only have 3 days left until the deadline (or 4, depends on how you count)), or as a BigUint-like implementation where we keep the ints in a K map and we access them through hooks, and we would still need to wrap the int ID in a struct so we would still have the same issues as for the full Rust struct implementation.

However, I think that some of the leaders (Ilia or Xiaohong or both) said that adding the u256 type as a fake native type is an acceptable solution, so, while I dislike it, that's what I ended up doing.

> 
> The rest of the implementation looks fine for me.

**Review by yanliu18 — APPROVED** [review]

**Comment by yanliu18** [comment]
@virgil-serbanuta I agree, we should work on more urgent task then. Thanks for the explanation. 
Then I am OK to approve the changes. We would have to add a `//TODO`:  comments for future review.

**Pull request merged**